core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java

```diff
@@ -20,9 +20,19 @@
 package org.elasticsearch.common.inject;
 
 /**
+ * This interface can be added to a Module to spawn sub modules. DO NOT USE.
  *
+ * This is fundamentally broken.
+ * <ul>
+ * <li>If you have a plugin with multiple modules, return all the modules at once.</li>
+ * <li>If you are trying to make the implementation of a module "pluggable", don't do it.
+ * This is not extendable because custom implementations (using onModule) cannot be
+ * registered before spawnModules() is called.</li>
+ * </ul>
+ * @deprecated Create modules directly
  */
-public interface SpawnModules {
+@Deprecated
```

Comment:
> But its not removed here, only deprecated.

deprecation is not the right mechanism as IMO we're deprecating the wrong thing here. It'd be like me deprecating the `Engine` just because there's an issue with index recovery. We need a separate issue for this... and sure, label it for 2.0... but it's not a blocker for beta1.

> You just have sour grapes because we are telling you your code fucking sucks.

:)

It's not my code... it's our code. I don't take this discussion personal at all and I have no problem with anyone saying our code sucks (you know it oh too well). I believe when one points to code that sucks it's awesome and it's even more awesome to also provide/suggest a better alternative. And yes.. I do find the discussion around this issue awesome by itself, regardless of what ends up being pushed... simply because I know it eventually leads to a better code base.

This PR is not a defining moment in anyone's life. Like you, I have my opinions when it comes to code. Sometimes it's aligned with others, sometimes it's not, whatever it is, we always find a way to agree and move forward, simply because we all come with good intentions to improve things. You're the same... I don't expect you (or anyone working on elasticsearch) to be different - and that, my dear friend, is what makes you awesome. We discuss and we resolve thing by agreeing to agree, not by agreeing to disagree - that's a monumental pillar in our culture.

Bottom line, this PR introduces changes I'm -1 on. These changes I'm "-1'ing" are not blocking the actual bug fix. I suggest to remove the controversial part of the PR (which... again... does not block the bug fix) and open a separate issue to discuss the suggested "cleanup".

Lets see what others have to say about it... and act on additional feedback.